### PR TITLE
factors: Support old WASI -rc interfaces

### PR DESCRIPTION
**Pull request by lann**
Also add a basic OutboundHttpFactor test.

### CODE REVIEW

File: crates/factor-outbound-http/src/wasi_2023_10_18.rs
Comment (lann) [inline crates/factor-outbound-http/src/wasi_2023_10_18.rs:1]: All these `wasi_2023_*` files are lightly adapted from the matching spin-core impls.

File: crates/factor-outbound-http/src/lib.rs [hunk 1]
```diff
@@ -1,5 +1,7 @@
 mod spin;
 mod wasi;
+mod wasi_2023_10_18;
+mod wasi_2023_11_10;
```
Comment (rylev) [inline crates/factor-outbound-http/src/lib.rs:4]: Why do we need these in this crate as opposed to using the definitions in `spin_world`?

**Review by rylev — COMMENTED** [review]
I'm not sure I fully understand the change here (i.e., why the bindings adapters need to be copied over). Could you provide more context on that?

File: crates/factor-outbound-http/src/lib.rs (hunk 1 above)
Comment (lann) [reply crates/factor-outbound-http/src/lib.rs:4]: These adapters need customized bindgen options to be compatible with the base `wasmtime-wasi{,-http}` impls.

(The _real_ answer is that I just blindly adapted @alexcrichton's code from `spin-core`. :sweat_smile:)

File: crates/factor-outbound-http/tests/factor_test.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    let req = Request::get("https://denied.test").body(Default::default())?;
+    let res = wasi_http.send_request(req, test_request_config());
+    let Err(err) = res else {
+        bail!("expected Err, got Ok");
+    };
+    assert!(matches!(err.downcast()?, ErrorCode::HttpRequestDenied));
+    Ok(())
+}
+
+fn test_request_config() -> OutgoingRequestConfig {
```
Comment (rylev) [inline crates/factor-outbound-http/tests/factor_test.rs:52]: FYI: this could be a `const` instead of a function

**Review by rylev — APPROVED** [review]

**Pull request merged**